File: BinarySearch/Python/binary_search.py
"""
Given a sorted array arr[] of n elements, write a function to search a given element x in arr[].
"""

def binary_search(ordered_list, l, r, x):
	"""
	Args:
    	ordered_lists (list): Ordered list of numbers
    	l (int): leftmost index
    	r (int): rightmost index
		x (int): number
	Returns:
    	Returns index of x in ordered_list if present, else -1 
	"""
	# Base case
	if r>=l: # Array has at least 1 element
		mid_point = (l + r) // 2

		if ordered_list[mid_point] == x:
			return mid_point
		elif ordered_list[mid_point] < x:
			return binary_search(ordered_list, mid_point + 1, r, x)
		else: # ordered_list[mid_point] > x
			return binary_search(ordered_list, l, mid_point, x)

	else:  # Array is empty
		return -1


def main():
	# Test 1
	array = [1,2,13,49,78,99]
	l = 0
	r = len(array) - 1
	x = 13

	index_test1 = binary_search(array, l, r, x)
	if index_test1 == 2:
		print("Your algo works!!!")
	else: 
		print("Your algo sucks!!!")

	# Test 2
	l = 0
	r = len(array) - 1
	x = 500
	index_test2 = binary_search(array, l, r, x)
	if index_test2 == -1:
		print("Your algo works!!!")
	else: 
		print("Your algo sucks!!!")


if __name__=='__main__':
	main()

File: BinarySearch/Scala/BinarySearch.scala
// Scala implementation of Binary Search algorithm

object BinarySearch {
	def binarySearch(orderedArray: Array[Int], 
					left: Int, 
					right: Int, 
					x: Int): Int = {

		// Base case
		if (right>=left) {
			
			var midPoint = left + (right - left)/2
			
			if (orderedArray(midPoint) == x)
				return midPoint
			
			if (x < orderedArray(midPoint))
				// x is in he leftmost part of the array
				return binarySearch(orderedArray, left, midPoint -1, x)
			 
			// orderedArray[midPoint] < x
			// x is in the rightmost part of the array
			return binarySearch(orderedArray, midPoint + 1, right, x)
		} else {
			return -1
		}
	}

	def main(args: Array[String]) {

		val array1 = Array(1, 2, 11, 33, 57, 99)
		val l1 = 0
		val r1 = array1.length
		val x1 = 2

		var index1 = binarySearch(array1, l1, r1 - 1, x1)

		val array2 = Array(1, 6, 9, 500, 1000)
		val l2 = 0
		val r2 = array2.length
		val x2 = 5000

		var index2 = binarySearch(array2, l2, r2 - 1, x2)


		if (index1 == 1)
			print("Your algo rocks! :)")
		else
			print("It does not work! :(")

		if (index2 == -1)
			print("Your algo rocks again! :)")
		else
			print("It still does not work! :(")
	}
}